# ENTREGABLE 4

# INSTRUCCIONES

Utilizar el archivo CSV (`dataset_banco_clean.csv`) con 45189 filas y 17 columnas y aplicar las técnicas de normalización del entregable 3.

In [39]:
# imports
from sklearn.datasets import load_iris
import numpy as np
import pandas as pd
from sklearn.preprocessing import MinMaxScaler
import plotly.express as px


In [40]:
ruta = "/content/dataset_banco_clean (2).csv"
data = pd.read_csv(ruta)

# Objetivo

Generar un model de clasificación capaz de predecir la clase de flor en función de las carácterísticas del dataset

* Aplicar las técnicas oportunas de procesamiento de datos

* Generar split de los datos

* Valorar diferentes modelos de clasificación

* Comparación entre modelos

* Ensemble

* Métricas

* Conclusiones finales

In [41]:
## A completar
data

age           job   marital  education default  balance housing loan  \
0       58    management   married   tertiary      no   2143.0     yes   no   
1       44    technician    single  secondary      no     29.0     yes   no   
2       33  entrepreneur   married  secondary      no      2.0     yes  yes   
3       47   blue-collar   married    unknown      no   1506.0     yes   no   
4       33       unknown    single    unknown      no      1.0      no   no   
...    ...           ...       ...        ...     ...      ...     ...  ...   
45184   51    technician   married   tertiary      no    825.0      no   no   
45185   71       retired  divorced    primary      no   1729.0      no   no   
45186   72       retired   married  secondary      no   5715.0      no   no   
45187   57   blue-collar   married  secondary      no    668.0      no   no   
45188   37  entrepreneur   married  secondary      no   2971.0      no   no   

         contact  day month  duration  campaign  pdays  previous poutcome    y  
0        unknown    5   may     261.0         1   -1.0         0  unknown   no  
1        unknown    5   may     151.0         1   -1.0         0  unknown   no  
2        unknown    5   may      76.0         1   -1.0         0  unknown   no  
3        unknown    5   may      92.0         1   -1.0         0  unknown   no  
4        unknown    5   may     198.0         1   -1.0         0  unknown   no  
...          ...  ...   ...       ...       ...    ...       ...      ...  ...  
45184   cellular   17   nov     977.0         3   -1.0         0  unknown  yes  
45185   cellular   17   nov     456.0         2   -1.0         0  unknown  yes  
45186   cellular   17   nov    1127.0         5  184.0         3  success  yes  
45187  telephone   17   nov     508.0         4   -1.0         0  unknown   no  
45188   cellular   17   nov     361.0         2  188.0        11    other   no  

[45189 rows x 17 columns]

In [42]:
data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45189 entries, 0 to 45188
Data columns (total 17 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   age        45189 non-null  int64  
 1   job        45189 non-null  object 
 2   marital    45189 non-null  object 
 3   education  45189 non-null  object 
 4   default    45189 non-null  object 
 5   balance    45189 non-null  float64
 6   housing    45189 non-null  object 
 7   loan       45189 non-null  object 
 8   contact    45189 non-null  object 
 9   day        45189 non-null  int64  
 10  month      45189 non-null  object 
 11  duration   45189 non-null  float64
 12  campaign   45189 non-null  int64  
 13  pdays      45189 non-null  float64
 14  previous   45189 non-null  int64  
 15  poutcome   45189 non-null  object 
 16  y          45189 non-null  object 
dtypes: float64(3), int64(4), object(10)
memory usage: 5.9+ MB


In [43]:
from sklearn.preprocessing import LabelEncoder
import pandas as pd

# Crear una instancia de LabelEncoder
label_encoder = LabelEncoder()
columnas_categoricas= data.select_dtypes(include=['object'])

# Iterar sobre cada columna categórica
for columna in columnas_categoricas:
    # Ajustar y transformar la columna
    data[columna] = label_encoder.fit_transform(data[columna])

    # Convertir a tipo float
    data[columna] = data[columna].astype(float)

data

age   job  marital  education  default  balance  housing  loan  \
0       58   4.0      1.0        2.0      0.0   2143.0      1.0   0.0   
1       44   9.0      2.0        1.0      0.0     29.0      1.0   0.0   
2       33   2.0      1.0        1.0      0.0      2.0      1.0   1.0   
3       47   1.0      1.0        3.0      0.0   1506.0      1.0   0.0   
4       33  11.0      2.0        3.0      0.0      1.0      0.0   0.0   
...    ...   ...      ...        ...      ...      ...      ...   ...   
45184   51   9.0      1.0        2.0      0.0    825.0      0.0   0.0   
45185   71   5.0      0.0        0.0      0.0   1729.0      0.0   0.0   
45186   72   5.0      1.0        1.0      0.0   5715.0      0.0   0.0   
45187   57   1.0      1.0        1.0      0.0    668.0      0.0   0.0   
45188   37   2.0      1.0        1.0      0.0   2971.0      0.0   0.0   

       contact  day  month  duration  campaign  pdays  previous  poutcome    y  
0          2.0    5    8.0     261.0         1   -1.0         0       3.0  0.0  
1          2.0    5    8.0     151.0         1   -1.0         0       3.0  0.0  
2          2.0    5    8.0      76.0         1   -1.0         0       3.0  0.0  
3          2.0    5    8.0      92.0         1   -1.0         0       3.0  0.0  
4          2.0    5    8.0     198.0         1   -1.0         0       3.0  0.0  
...        ...  ...    ...       ...       ...    ...       ...       ...  ...  
45184      0.0   17    9.0     977.0         3   -1.0         0       3.0  1.0  
45185      0.0   17    9.0     456.0         2   -1.0         0       3.0  1.0  
45186      0.0   17    9.0    1127.0         5  184.0         3       2.0  1.0  
45187      1.0   17    9.0     508.0         4   -1.0         0       3.0  0.0  
45188      0.0   17    9.0     361.0         2  188.0        11       1.0  0.0  

[45189 rows x 17 columns]

In [44]:
for col in data.columns[0:15]:
    # Calcular valores mínimo y máximo de la columna
    minimo = data[col].min()
    maximo = data[col].max()
    print(f'Máximo/mínimo columna "{col}": {maximo:.1f}/{minimo:.1f}')

Máximo/mínimo columna "age": 95.0/18.0
Máximo/mínimo columna "job": 11.0/0.0
Máximo/mínimo columna "marital": 2.0/0.0
Máximo/mínimo columna "education": 3.0/0.0
Máximo/mínimo columna "default": 1.0/0.0
Máximo/mínimo columna "balance": 527532.0/-8019.0
Máximo/mínimo columna "housing": 1.0/0.0
Máximo/mínimo columna "loan": 1.0/0.0
Máximo/mínimo columna "contact": 2.0/0.0
Máximo/mínimo columna "day": 31.0/1.0
Máximo/mínimo columna "month": 11.0/0.0
Máximo/mínimo columna "duration": 4918.0/1.0
Máximo/mínimo columna "campaign": 63.0/1.0
Máximo/mínimo columna "pdays": 871.0/-1.0
Máximo/mínimo columna "previous": 58.0/0.0


In [45]:
data['y'].value_counts(normalize=True)

y
0.0    0.883047
1.0    0.116953
Name: proportion, dtype: float64

In [46]:
x=data.drop(columns=['y'])
y=data['y']

In [47]:
from sklearn.model_selection import train_test_split

## A COMPLETAR

'''Crear los sets de entrenamiento (70%), validación (15%)
    y prueba (15%)'''

def crear_train_validation_test (x,y):

  xtr, x_rest, ytr, y_rest = train_test_split(x, y, test_size=0.3, random_state=43)
  xvl, xts, yvl, yts = train_test_split(x_rest, y_rest, test_size=0.5, random_state=43)

  return xtr, ytr, xvl, yvl, xts, yts


# Ejecutar la función

x_train, y_train, x_val, y_val, x_test, y_test = crear_train_validation_test(x,y)

# Verificar
print(x_train.shape, y_train.shape)
print(x_val.shape, y_val.shape)
print(x_test.shape, y_test.shape)

(31632, 16) (31632,)
(6778, 16) (6778,)
(6779, 16) (6779,)


In [48]:
def escalar_datos(xtr, xvl, xts):
    '''Escalar datos de entrada (entrenamiento, validación y prueba)
    en el rango de 0 a 1'''
    scaler = MinMaxScaler()

    # Fit + transform sobre el set de entrenamiento
    xtr_s = scaler.fit_transform(xtr)

    # Transform sobre prueba y validación
    xvl_s = scaler.transform(xvl)
    xts_s = scaler.transform(xts)

    return xtr_s, xvl_s, xts_s

# Ejecutar la función
x_train_s, x_val_s, x_test_s = escalar_datos(x_train, x_val, x_test)

# Verificar
print(x_train_s.min(axis=0),x_train_s.max(axis=0))
print(x_val_s.min(axis=0),x_val_s.max(axis=0))
print(x_test_s.min(axis=0),x_test_s.max(axis=0))

[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.] [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
[0.01298701 0.         0.         0.         0.         0.01101109
 0.         0.         0.         0.         0.         0.00020338
 0.         0.         0.         0.        ] [0.97402597 1.         1.         1.         1.         0.1395572
 1.         1.         1.         1.         1.         0.78909904
 0.85964912 1.00470035 1.05454545 1.        ]
[0.         0.         0.         0.         0.         0.00867704
 0.         0.         0.         0.         0.         0.
 0.         0.         0.         0.        ] [0.98701299 1.         1.         1.         1.         0.14789815
 1.         1.         1.         1.         1.         0.56416514
 1.0877193  1.02467685 0.52727273 1.        ]


#Modelos

Árbol de clasificación:

In [49]:
## A COMPLETAR

from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import f1_score

def entrenar_y_evaluar_arbol(x_train, y_train, x_val, y_val, max_depth=None):
    '''Entrena y evalúa un árbol de clasificación con los datos proporcionados'''
    # Inicializar el clasificador de árbol de decisión
    clf = DecisionTreeClassifier(max_depth=max_depth, random_state=42)

    # Entrenar el modelo con los datos de entrenamiento
    clf.fit(x_train, y_train)

    # Predecir las etiquetas de clase en los datos de validación
    y_pred_val = clf.predict(x_val)

    # Calcular la precisión en los datos de validación
    accuracy = accuracy_score(y_val, y_pred_val)
    puntaje_f1 = f1_score(y_val, y_pred_val, average='weighted')  # 'macro' es útil cuando tienes clases desbalanceadas

    return clf, accuracy, puntaje_f1


# Entrenar y evaluar un árbol de clasificación
arbol, precision, puntaje_f1 = entrenar_y_evaluar_arbol(x_train_s, y_train, x_val_s, y_val)

# Imprimir la precisión obtenida
print("Precisión del árbol de clasificación:", precision)
print("Precisión del árbol de clasificación f1_score:", puntaje_f1)



Precisión del árbol de clasificación: 0.8855119504278548
Precisión del árbol de clasificación f1_score: 0.8866121379380721


Clasificador kNN:

In [50]:
## A COMPLETAR

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

def entrenar_y_evaluar_knn(x_train, y_train, x_val, y_val, n_neighbors=5):
    '''Entrena y evalúa un clasificador kNN con los datos proporcionados'''
    # Inicializar el clasificador kNN
    clf = KNeighborsClassifier(n_neighbors=n_neighbors)

    # Entrenar el modelo con los datos de entrenamiento
    clf.fit(x_train, y_train)

    # Predecir las etiquetas de clase en los datos de validación
    y_pred_val = clf.predict(x_val)

    # Calcular la precisión en los datos de validación
    accuracy = accuracy_score(y_val, y_pred_val)
    puntaje_f1 = f1_score(y_val, y_pred_val, average='weighted')

    return clf, accuracy, puntaje_f1

# Entrenar y evaluar un clasificador kNN
knn, precision_knn, puntaje_f1 = entrenar_y_evaluar_knn(x_train_s, y_train, x_val_s, y_val)

# Imprimir la precisión obtenida
print("Precisión del clasificador kNN:", precision_knn)
print("Precisión kNN f1_score:", puntaje_f1)

Precisión del clasificador kNN: 0.8930362938920036
Precisión kNN f1_score: 0.8728202995821841


SVM

In [51]:
from sklearn.svm import SVC

def entrenar_y_evaluar_svm(x_train, y_train, x_val, y_val, kernel='rbf', C=1.0):
    '''Entrena y evalúa un modelo de SVM con los datos proporcionados'''
    # Inicializar el clasificador SVM
    clf = SVC(kernel=kernel, C=C, random_state=42)

    # Entrenar el modelo con los datos de entrenamiento
    clf.fit(x_train, y_train)

    # Predecir las etiquetas de clase en los datos de validación
    y_pred_val = clf.predict(x_val)

    # Calcular la precisión en los datos de validación
    accuracy = accuracy_score(y_val, y_pred_val)
    puntaje_f1 = f1_score(y_val, y_pred_val, average='weighted')

    return clf, accuracy, puntaje_f1


# Entrenar y evaluar un modelo SVM con escalamiento
svm_con_escalamiento, precision_svm_con_escalamiento, puntaje_f1 = entrenar_y_evaluar_svm(x_train_s, y_train, x_val_s, y_val)

# Imprimir la precisión obtenida
print("Precisión del modelo SVM con escalamiento:", precision_svm_con_escalamiento)
print("Precisión del modelo SVM con f1score:", puntaje_f1)

Precisión del modelo SVM con escalamiento: 0.8937739746237828
Precisión del modelo SVM con f1score: 0.8645877083834685


Naive Bayes

In [52]:
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score

def entrenar_y_evaluar_naive_bayes(x_train, y_train, x_val, y_val):
    '''Entrena y evalúa un modelo de Naive Bayes con los datos proporcionados'''
    # Inicializar el clasificador Naive Bayes
    clf = GaussianNB()

    # Entrenar el modelo con los datos de entrenamiento
    clf.fit(x_train, y_train)

    # Predecir las etiquetas de clase en los datos de validación
    y_pred_val = clf.predict(x_val)

    # Calcular la precisión en los datos de validación
    accuracy = accuracy_score(y_val, y_pred_val)
    puntaje_f1 = f1_score(y_val, y_pred_val, average='weighted')

    return clf, accuracy, puntaje_f1

# Entrenar y evaluar un modelo Naive Bayes con escalamiento
naive_bayes_con_escalamiento, precision_naive_bayes_con_escalamiento, puntaje_f1 = entrenar_y_evaluar_naive_bayes(x_train_s, y_train, x_val_s, y_val)

# Imprimir la precisión obtenida
print("Precisión del modelo Naive Bayes con escalamiento:", precision_naive_bayes_con_escalamiento)
print("Precisión del modelo Naive Bayes con f1score:", puntaje_f1)

Precisión del modelo Naive Bayes con escalamiento: 0.830628503983476
Precisión del modelo Naive Bayes con f1score: 0.844971870874554


Random Forest

In [53]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

def entrenar_y_evaluar_random_forest(x_train, y_train, x_val, y_val, n_estimators=100, max_depth=None, random_state=42):
    '''Entrena y evalúa un modelo de Random Forest con los datos proporcionados'''
    # Inicializar el clasificador Random Forest
    clf = RandomForestClassifier(n_estimators=n_estimators, max_depth=max_depth, random_state=43)

    # Entrenar el modelo con los datos de entrenamiento
    clf.fit(x_train, y_train)

    # Predecir las etiquetas de clase en los datos de validación
    y_pred_val = clf.predict(x_val)

    # Calcular la precisión en los datos de validación
    accuracy = accuracy_score(y_val, y_pred_val)
    puntaje_f1 = f1_score(y_val, y_pred_val, average='weighted')

    return clf, accuracy, puntaje_f1

# Entrenar y evaluar un modelo Random Forest con escalamiento
random_forest_con_escalamiento, precision_random_forest_con_escalamiento, puntaje_f1 = entrenar_y_evaluar_random_forest(x_train_s, y_train, x_val_s, y_val)

# Imprimir la precisión obtenida
print("Precisión del modelo Random Forest con escalamiento:", precision_random_forest_con_escalamiento)
print("Precisión del modelo Random Forest con f1score:", puntaje_f1 )


Precisión del modelo Random Forest con escalamiento: 0.9089701976984361
Precisión del modelo Random Forest con f1score: 0.9014387939596982


#ENSEMBLES

In [38]:
from sklearn.ensemble import VotingClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, f1_score

def train_and_evaluate_classifier(clf, x_train, y_train, x_val, y_val):
    '''Entrena y evalúa un clasificador con los datos proporcionados'''
    # Entrenar el clasificador
    clf.fit(x_train, y_train)

    # Predecir las etiquetas de clase en los datos de validación
    y_pred_val = clf.predict(x_val)

    # Calcular la precisión en los datos de validación
    accuracy = accuracy_score(y_val, y_pred_val)

    # Calcular el puntaje F1 en los datos de validación
    f1 = f1_score(y_val, y_pred_val, average='weighted')

    return clf, accuracy, f1

# Crear los clasificadores individuales
knn_classifier = KNeighborsClassifier()
svm_classifier = SVC(probability=True)  # Habilita las probabilidades para VotingClassifier
nb_classifier = GaussianNB()
rf_classifier = RandomForestClassifier(n_estimators=100, max_depth=None, random_state=42)

# Crear el VotingClassifier con los modelos específicos
ensemble = VotingClassifier(estimators=[
    ('knn', knn_classifier),
    ('svm', svm_classifier),
    ('nb', nb_classifier),
    ('rf', rf_classifier)
], voting='hard')  # Cambia a 'soft' si deseas utilizar votación suave

# Entrenar y evaluar el ensemble
ensemble, precision_ensemble, f1_ensemble = train_and_evaluate_classifier(ensemble, x_train_s, y_train, x_val_s, y_val)

# Imprimir la precisión y el puntaje F1 obtenidos
print("Precisión del modelo Voting Classifier:", precision_ensemble)
print("Puntaje F1 del modelo Voting Classifier:", f1_ensemble)


Precisión del modelo Voting Classifier: 0.8967246975509
Puntaje F1 del modelo Voting Classifier: 0.8694947461702054


# Conclusiones

En conclusión, una vez obtenido el dataset limpio, se realiza la parte de normalización para contar con valores entre 0 y 1 de las variables para con ello poder realizar los modelos de predicción.

Además, se pudo evidenciar que el modelo random forest con 0.90% en la métrica de f1 score, es el modelo que mayor precisión tiene al momento de determinar los clientes contaran con una poliza en el banco.

Teniendo en cuenta que los demás modelos arrojan un puntaje alrededor del 80% lo cual se puede considerar efectivo dentro de las predicciones.
